### PR TITLE
sso: support multiple identity providers

### PR DESCRIPTION
**Pull request by jphines**
## Problem

We've been walking down the path to support multiple identity providers in several previous iterations. This last remaining stand is how to support _configuration_ for multiple and different identity providers.  The existing configuration structure is pretty inflexible and rigid when it comes to defining configuration, especially given our constraint to use environment variables.

We previously introduced `viper` to help solve some of these problems, but viper proved frustrating and lacking enough opinion to operate within. We roll that solution back here and instead use https://micro.mu/docs/go-config.html

## Solution

We introduced a new configuration component based on https://micro.mu/docs/go-config.html to build this new, more complicated configuration mechanism.  This is a ground-up re-write of the configuration mechanism we use, found separately in `configuration.go`.

## Notes

This now looks like:
```
### session
SESSION COOKIE_NAME
SESSION_COOKIE_SECRET
SESSION_COOKIE_EXPIRE
SESSION_COOKIE_DOMAIN
SESSION_COOKIE_REFRESH
SESSION_COOKIE_SECURE
SESSION_COOKIE_HTTPONLY
SESSION_LIFETIME
SESSION_KEY


### client
CLIENT_PROXY_ID
CLIENT_PROXY_SECRET


### provider config for google
PROVIDER_*_TYPE
PROVIDER_*_SLUG
PROVIDER_*_CLIENT_ID
PROVIDER_*_CLIENT_SECRET
PROVIDER_*_SCOPE

### google specific 
PROVIDER_*_GOOGLE_CREDENTIALS
PROVIDER_*_GOOGLE_IMPERSONATE

### okta specific
PROVIDER_*_OKTA_URL
PROVIDER_*_OKTA_SERVER

### group refresh
PROVIDER_*_GROUPCACHE_INTERVAL_REFRESH
PROVIDER_*_GROUPCACHE_INTERVAL_PROVIDER


# server config
SERVER_SCHEME
SERVER_HOST
SERVER_PORT
SERVER_TIMEOUT_REQUEST
SERVER_TIMEOUT_WRITE
SERVER_TIMEOUT_READ


# authorize config
AUTHORIZE_PROXY_DOMAINS
AUTHORIZE_EMAIL_DOMAINS
AUTHORIZE_EMAIL_ADDRESSES


# metrics config 
METRICS_STATSD_PORT
METRICS_STATSD_HOST


# logging config
LOGGING_ENABLE
LOGGING_LEVEL
```

### CODE REVIEW

File: internal/auth/mux.go [hunk 1]
```diff
@@ -17,34 +16,33 @@ type AuthenticatorMux struct {
...
+			return nil, err
+		}
+
 		idpSlug := idp.Data().ProviderSlug
-		authenticator, err := NewAuthenticator(opts,
+		authenticator, err := NewAuthenticator(config,
 			SetValidator(validator),
 			SetProvider(idp),
-			SetCookieStore(opts, idpSlug),
+			SetCookieStore(config, idpSlug),
```
Comment (Jusshersmith) [inline internal/auth/mux.go]: Should we be even more explicit here with `SetCookieStore` and pass in `config.SessionConfig` instead? It doesn't look like it needs to access any other config data structures

File: internal/auth/configuration.go [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+			Level:  "info",
+		},
+		MetricsConfig: MetricsConfig{
+			StatsdConfig: StatsdConfig{
+				Port: 8125,
+				Host: "localhost",
+			},
+		},
+		// we provide no defaults for these right now
+		AuthorizeConfig: AuthorizeConfig{
```
Comment (Jusshersmith) [inline internal/auth/configuration.go]: I'm not entirely sure what's best for this name - so I don't have strong opinions, but wanted to throw it out there: would `AuthenticatorConfig` be clearer? I'm largely basing that off where it's passed into though.

Comment (Jusshersmith) [reply internal/auth/configuration.go]: Hmm, after looking through again that doesn't seem right!

File: internal/auth/configuration.go [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+	if err := pc.GroupCacheConfig.Validate(); err != nil {
+		return xerrors.Errorf("invalid provider.groupcache config: %w", err)
+	}
+
+	return nil
+}
+
+type GoogleProviderConfig struct {
+	Credentials    string `mapstructure:"credentials"`
+	Impersonate    string `mapstructure:"impersonate"`
```
Comment (Jusshersmith) [inline internal/auth/configuration.go]: `UserToImpersonate` / `ImpersonateUser`? 🤷‍♂

File: internal/auth/options.go [hunk 4]
```diff
@@ -3,310 +3,64 @@ package auth
...
 	"github.com/buzzfeed/sso/internal/auth/providers"
+	"github.com/buzzfeed/sso/internal/pkg/aead"
 	"github.com/buzzfeed/sso/internal/pkg/groups"
+	"github.com/buzzfeed/sso/internal/pkg/sessions"
 
 	"github.com/datadog/datadog-go/statsd"
-	"github.com/spf13/viper"
 )
 
-// Options are config options that can be set by environment variables
```
Comment (Jusshersmith) [inline internal/auth/options.go]: I wonder what the best way is for us to capture this kind of information now? Could using something like godoc help with this, among other things?

File: internal/auth/configuration.go [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+	ApprovalPrompt string `mapstructure:"prompt"`
+}
+
+func (gpc GoogleProviderConfig) Validate() error {
+	// TOOD: what can we validate here?
+	return nil
```
Comment (Jusshersmith) [inline internal/auth/configuration.go]: For the Okta provider, it seems like we actually test for the `OrgURL` and `ServerID` both here _and_ in the actual provider itself. 

In a similar fashion, the Google provider tests for `adminEmail` and `credsFilePath` - I'm not sure if we want to have it just be tested in once place, or in both places? (But, I guess we should be consistent with both providers with whichever direction we go)

File: internal/auth/configuration.go [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+	if cookie.String() == "" {
+		return xerrors.Errorf("invalid cookie.name: %q", cc.Name)
+	}
+
+	if cc.Secret == "" {
+		return xerrors.New("no cookie.secret configured")
+	}
+
+	if _, err := base64.StdEncoding.DecodeString(cc.Secret); err != nil {
+		return xerrors.Errorf("invalid cookie.secret: %w", err)
```
Comment (Jusshersmith) [inline internal/auth/configuration.go]: Upon error we previously would also output how this variable _should_ be created - I don't think it necessarily needs to be added to the output, but I just want to make sure that information isn't lost as I know it's something that was brought up in the past.

File: internal/auth/configuration.go (hunk 5 above)
Comment (jphines) [reply internal/auth/configuration.go]: For Okta, those variable are mandatory. For Google, they are optional.

**Comment by Jusshersmith** [comment]
Below is a list of `OLD_CONFIG` -> `NEW_CONFIG` pairs to further aid and support anyone moving over to this new configuration set.
```
### SESSION
(NEW)                -> SESSION_COOKIE_NAME
COOKIE_SECRET        -> SESSION_COOKIE_SECRET
COOKIE_EXPIRE        -> SESSION_COOKIE_EXPIRE
COOKIE_DOMAIN        -> SESSION_COOKIE_DOMAIN
COOKIE_REFRESH       -> SESSION_COOKIE_REFRESH
COOKIE_SECURE        -> SESSION_COOKIE_SECURE
COOKIE_HTTP_ONLY     -> SESSION_COOKIE_HTTPONLY
SESSION_LIFETIME_TTL -> SESSION_LIFETIME
AUTH_CODE_SECRET     -> SESSION_KEY


### CLIENT
PROXY_CLIENT_ID     -> CLIENT_PROXY_ID
PROXY_CLIENT_SECRET -> CLIENT_PROXY_SECRET


### PROVIDER CONFIG FOR GOOGLE
(NEW)         -> PROVIDER_*_TYPE
(NEW)         -> PROVIDER_*_SLUG
CLIENT_ID     -> PROVIDER_*_CLIENT_ID
CLIENT_SECRET -> PROVIDER_*_CLIENT_SECRET
SCOPE         -> PROVIDER_*_SCOPE

### GOOGLE SPECIFIC 
GOOGLE_SERVICE_ACCOUNT_JSON -> PROVIDER_*_GOOGLE_CREDENTIALS
GOOGLE_ADMIN_EMAIL          -> PROVIDER_*_GOOGLE_IMPERSONATE

### OKTA SPECIFIC
OKTA_ORG_URL       -> PROVIDER_*_OKTA_URL
PROVIDER_SERVER_ID -> PROVIDER_*_OKTA_SERVER

### GROUP REFRESH
GROUPS_CACHE_REFRESH_TTL  -> PROVIDER_*_GROUPCACHE_INTERVAL_REFRESH
GROUPS_CACHE_PROVIDER_TTL -> PROVIDER_*_GROUPCACHE_INTERVAL_PROVIDER


# SERVER CONFIG
(NEW)             -> SERVER_SCHEME
HOST              -> SERVER_HOST
PORT              -> SERVER_PORT
REQUEST_TIMEOUT   -> SERVER_TIMEOUT_REQUEST
TCP_WRITE_TIMEOUT -> SERVER_TIMEOUT_WRITE
TCP_READ_TIMEOUT  -> SERVER_TIMEOUT_READ


# AUTHORIZE CONFIG
PROXY_ROOT_DOMAIN   -> AUTHORIZE_PROXY_DOMAINS
SSO_EMAIL_DOMAIN    -> AUTHORIZE_EMAIL_DOMAINS
SSO_EMAIL_ADDRESSES -> AUTHORIZE_EMAIL_ADDRESSES


# METRICS CONFIG 
STATSD_PORT -> METRICS_STATSD_PORT
STATSD_HOST -> METRICS_STATSD_HOST


# LOGGING CONFIG
REQUSEST_LOGGING -> LOGGING_ENABLE
(NEW)            -> LOGGING_LEVEL
```

**Pull request merged**